src/Microsoft.OData.Client/Serialization/HttpWebRequestMessage.cs

```diff
@@ -63,149 +63,143 @@ public class HttpWebRequestMessage : DataServiceClientRequestMessage
         /// Initializes the <see cref="HttpWebRequestMessage"/> class.
         /// </summary>
-        static HttpWebRequestMessage()
+        public HttpWebRequestMessage(DataServiceClientRequestMessageArgs args) 
+            : base(args.ActualMethod)
         {
-            IsRunningOnSilverlight = typeof(System.Windows.Input.ICommand).GetAssembly().GetType("System.Net.Browser.WebRequestCreator", false) != null;
+            _requestMessage = new HttpRequestMessage();
+            _messageStream = new MemoryStream();
+            _handler = new HttpClientHandler();
+            _client = new HttpClient(_handler, disposeHandler: true);
+            _contentHeaderValueCache = new Dictionary<string, string>();
 
-            PropertyInfo pi = typeof(HttpWebRequest).GetProperty("ContentLength");
-            if (pi != null)
-            {
-                ContentLengthMethodSetter = pi.GetSetMethod();
-            }
+            _effectiveHttpMethod = args.Method;
+            _requestUrl = args.RequestUri;
 
-            pi = typeof(HttpWebRequest).GetProperty("UserAgent");
-            if (pi != null)
+            _requestMessage = new HttpRequestMessage(new HttpMethod(this.ActualMethod), _requestUrl);
+
+            // Now set the headers.
+            foreach (var keyValue in args.Headers)
             {
-                UserAgentMethodSetter = pi.GetSetMethod();
+                this.SetHeader(keyValue.Key, keyValue.Value);
             }
         }
-#endif
 
         /// <summary>
-        /// Creates a new instance of HttpWebRequestMessage.
+        /// Returns the collection of request headers.
         /// </summary>
-        /// <param name="args">Arguments for creating the request message.</param>
-        [SuppressMessage("Microsoft.Usage", "CA2214:DoNotCallOverridableMethodsInConstructors", Justification = "SetHeader is a safe virtual method to be called from the constructor")]
-        public HttpWebRequestMessage(DataServiceClientRequestMessageArgs args)
-            : base(args.ActualMethod)
+        public override IEnumerable<KeyValuePair<string, string>> Headers
         {
-            Util.CheckArgumentNull(args, "args");
-            Debug.Assert(args.RequestUri.IsAbsoluteUri, "request uri is not absolute uri");
-            Debug.Assert(
-                args.RequestUri.Scheme.Equals("http", StringComparison.OrdinalIgnoreCase) ||
-                args.RequestUri.Scheme.Equals("https", StringComparison.OrdinalIgnoreCase),
-                "request uri is not for HTTP");
-            this.effectiveHttpMethod = args.Method;
-            this.requestUrl = args.RequestUri;
-
-            this.httpRequest = HttpWebRequestMessage.CreateRequest(this.ActualMethod, this.Url, args);
-
-            // Now set the headers.
-            foreach (var keyValue in args.Headers)
+            get
             {
-                this.SetHeader(keyValue.Key, keyValue.Value);
+                if (_requestMessage.Content != null)

```

Comment:
@ElizabethOkerio If the headers are likely to change during the lifetime of the object, then is justifiable but if they don't then it'd make sense to use `Concat` first time and store the result into a private variable as opposed to calling `Concat` each time the `Headers` property is referenced

Reply:
so the headers get set once per request.